aidial_adapter_bedrock/utils/adapter_deployments.py

```diff
@@ -0,0 +1,144 @@
+    """
+    The deployment id of the corresponding Bedrock model.
+    The upstream request to the Bedrock service will use this deployment id.
+    """
+
+    reference_deployment_id: _D
+    """
+    The reference Bedrock deployment which is known to share
+    the same API as `upstream_deployment_id`.
+    """
+
+    @classmethod
+    def supported(
+        cls, *, deployment_id: str | None = None, upstream: _D
+    ) -> Self:
+        return cls(
+            adapter_deployment_id=deployment_id or upstream.value,
+            upstream_deployment_id=upstream.value,
+            reference_deployment_id=upstream,
+        )
+
+    def compat(self, deployment_id: str) -> "AdapterDeployment[_D]":
+        return AdapterDeployment(
+            adapter_deployment_id=deployment_id,
+            upstream_deployment_id=deployment_id,
+            reference_deployment_id=self.reference_deployment_id,
+        )
+
+    def clone(self, reference_deployment_id: _T) -> "AdapterDeployment[_T]":
+        return AdapterDeployment(
+            adapter_deployment_id=self.adapter_deployment_id,
+            upstream_deployment_id=self.upstream_deployment_id,
+            reference_deployment_id=reference_deployment_id,
+        )
+
+
+AdapterChatCompletionDeployment = AdapterDeployment[ChatCompletionDeployment]
+AdapterEmbeddingsDeployment = AdapterDeployment[EmbeddingsDeployment]
+
+
+class AdapterDeployments(BaseModel):
+    chat_completions: Dict[str, AdapterChatCompletionDeployment]
+    embeddings: Dict[str, AdapterEmbeddingsDeployment]
+
+    @classmethod
+    def create(cls, *, compat_mapping: Dict[str, str]) -> "AdapterDeployments":
+
+        chat_completions = {e.value for e in ChatCompletionDeployment}
+        embeddings = {e.value for e in EmbeddingsDeployment}
+
+        for deployment_id, supported_id in compat_mapping.items():
+            if deployment_id in chat_completions or deployment_id in embeddings:
+                log.warning(
+                    f"{deployment_id!r} is one of the Bedrock deployments supported by the adapter already. "
+                    f"Remove {deployment_id!r} from the compatibility mapping to avoid the warning."
+                )
+
+                if (
+                    deployment_id in chat_completions
+                    and supported_id in embeddings
+                ):
+                    raise ValueError(
+                        f"The chat completion deployment {deployment_id!r} is mapped onto the embeddings deployment {supported_id!r}"
+                    )
+
+                if (
+                    deployment_id in embeddings
+                    and supported_id in chat_completions
+                ):
+                    raise ValueError(
+                        f"The embeddings deployment {deployment_id!r} is mapped onto the chat completion deployment {supported_id!r}"
+                    )
+
+        chat_completions = _create_deployments(
+            compat_mapping,
```

Comment:
We use `.pop` in this operation, let's create copy of `compat_mapping` for this, to avoid this hidden modification of input parameter

Reply:
Fixed. Added dict copy to `_create_deployments` to avoid leaking implementation details.